core/unix/os.c

```diff
@@ -7444,26 +7445,37 @@ pre_system_call(dcontext_t *dcontext)
             size_t sizemask;
         } data_t;
         data_t *data = (data_t *)sys_param(dcontext, 5);
-        kernel_sigset_t *mask = (kernel_sigset_t *)data->sigmask;
-        size_t sizemask = (size_t)data->sizemask;
-        kernel_sigset_t nullsigmask;
-        memset(&nullsigmask, 0, sizeof(nullsigmask));
-        /* syscall ABI, struct is in the  app's stack */
-        safe_write_ex((void *)&data->sigmask, sizeof(kernel_sigset_t),
-                      (const void *)&nullsigmask, NULL);
+        kernel_sigset_t *mask;
+        if (!safe_read(&data->sigmask, sizeof(kernel_sigset_t), &mask)) {
+            LOG(THREAD, LOG_SYSCALLS, 2, "\treturning EFAULT to app for pselect6\n");
+            set_failure_return_val(dcontext, EINVAL);
+            DODEBUG({ dcontext->expect_last_syscall_to_fail = true; });
+        }
+        size_t sizemask;
+        if (!safe_read(&data->sizemask, sizeof(size_t), &sizemask)) {
```

Comment:
Simpler to declare a data_t and do one safe_read into it

Reply:
Done